ext/native/thin3d/VulkanRenderManager.cpp

```diff
@@ -0,0 +1,1328 @@
+			attachments[1].stencilLoadOp = VK_ATTACHMENT_LOAD_OP_LOAD;
+			break;
+		case VKRRenderPassAction::DONT_CARE:
+			attachments[1].loadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE;
+			attachments[1].stencilLoadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE;
+			break;
+		}
+		for (int color = 0; color < 3; color++) {
+			switch ((VKRRenderPassAction)color) {
+			case VKRRenderPassAction::CLEAR: attachments[0].loadOp = VK_ATTACHMENT_LOAD_OP_CLEAR; break;
+			case VKRRenderPassAction::KEEP: attachments[0].loadOp = VK_ATTACHMENT_LOAD_OP_LOAD; break;
+			case VKRRenderPassAction::DONT_CARE: attachments[0].loadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE; break;
+			}
+			int index = RPIndex((VKRRenderPassAction)color, (VKRRenderPassAction)depth);
+			vkCreateRenderPass(vulkan_->GetDevice(), &rp, nullptr, &renderPasses_[index]);
+		}
+	}
+}
+
+void VulkanRenderManager::Clear(uint32_t clearColor, float clearZ, int clearStencil, int clearMask) {
+	_dbg_assert_(G3D, curRenderStep_ && curRenderStep_->stepType == VKRStepType::RENDER);
+	// If this is the first drawing command, merge it into the pass.
+	if (curRenderStep_->render.numDraws == 0) {
+		curRenderStep_->render.clearColor = clearColor;
+		curRenderStep_->render.clearDepth = clearZ;
+		curRenderStep_->render.clearStencil = clearStencil;
+		curRenderStep_->render.color = (clearMask & VK_IMAGE_ASPECT_COLOR_BIT) ? VKRRenderPassAction::CLEAR : VKRRenderPassAction::KEEP;
+		curRenderStep_->render.depthStencil = (clearMask & (VK_IMAGE_ASPECT_DEPTH_BIT | VK_IMAGE_ASPECT_STENCIL_BIT)) ? VKRRenderPassAction::CLEAR : VKRRenderPassAction::KEEP;
+	} else {
+		VkRenderData data{ VKRRenderCommand::CLEAR };
+		data.clear.clearColor = clearColor;
+		data.clear.clearZ = clearZ;
+		data.clear.clearStencil = clearStencil;
+		data.clear.clearMask = clearMask;
+		curRenderStep_->commands.push_back(data);
+	}
+}
+
+void VulkanRenderManager::CopyFramebuffer(VKRFramebuffer *src, VkRect2D srcRect, VKRFramebuffer *dst, VkOffset2D dstPos, int aspectMask) {
+	_dbg_assert_msg_(G3D, srcRect.offset.x >= 0, "srcrect offset x < 0");
+	_dbg_assert_msg_(G3D, srcRect.offset.y >= 0, "srcrect offset y < 0");
+	_dbg_assert_msg_(G3D, srcRect.offset.x + srcRect.extent.width <= (uint32_t)src->width, "srcrect offset x + extent > width");
+	_dbg_assert_msg_(G3D, srcRect.offset.y + srcRect.extent.height <= (uint32_t)src->height, "srcrect offset y + extent > height");
+	_dbg_assert_msg_(G3D, dstPos.x >= 0, "dstPos offset x < 0");
+	_dbg_assert_msg_(G3D, dstPos.y >= 0, "dstPos offset y < 0");
+	_dbg_assert_msg_(G3D, dstPos.x + srcRect.extent.width <= (uint32_t)dst->width, "dstPos + extent x > width");
+	_dbg_assert_msg_(G3D, dstPos.y + srcRect.extent.height <= (uint32_t)dst->height, "dstPos + extent y > height");
+
+	VKRStep *step = new VKRStep{ VKRStepType::COPY };
+
+	step->copy.aspectMask = aspectMask;
+	step->copy.src = src;
+	step->copy.srcRect = srcRect;
+	step->copy.dst = dst;
+	step->copy.dstPos = dstPos;
+
+	std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
Hmm, not all pushes to steps_ have locks and I think this mutex_ is no longer used... just noting.

-[Unknown]